### PR TITLE
Improve undeclared goption key messages.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** user messages.

- [x] Added / updated test-suite
- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by Zimmi48** [comment]
Documentation update will be included in #11961 if this PR is merged first.

File: vernac/vernacentries.ml [hunk 1]
```diff
@@ -1523,26 +1523,34 @@ let vernac_set_option ~locality table v = match v with
       vernac_set_option0 ~locality table v
 | _ -> vernac_set_option0 ~locality table v
 
-let vernac_add_option key lv =
-  let f = function
-    | StringRefValue s -> (get_string_table key).add (Global.env()) s
-    | QualidRefValue locqid -> (get_ref_table key).add (Global.env()) locqid
+(* We have to request twice the same polymorphic function as argument
+   because OCaml won't compute the same type for it in both cases. *)
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml:1527]: We could alternatively use a record, I'll push an example commit

Also maybe iter_table should live in goptions?

**Review by SkySkimmer — APPROVED** [review]
Will wait for reaction to my comment/commit then merge.

File: vernac/vernacentries.ml (hunk 1 above)
Comment (Zimmi48) [reply vernac/vernacentries.ml]: Yes, I didn't like the repetition, your commit makes things much nicer.

Moving `iter_table` to `goptions` sounds relevant too, even if for now `vernacentries` would be the only user. I'll take care of doing the move.

Comment (Zimmi48) [reply vernac/vernacentries.ml]: Done. Let me know what you think of this last commit.

**Review by SkySkimmer — APPROVED** [review]

**Comment by Zimmi48** [comment]
Hum, I realize that this PR actually changes the API so might not be worth backporting.

**Comment by SkySkimmer** [comment]
I don't think we're worried about adding to the api when backporting.

**Comment by Zimmi48** [comment]
If all the PR did was to add to the API, it would be fine, but it also changes existing API (`vernac/vernacexpr.ml`). I was bitten once by this issue when I was the RM (I broke Coq-Elpi). Plugins should expect to stay compatible without effort with patch-level releases.

**Comment by Zimmi48** [comment]
I forgot to update the documentation. I'll fix this soon in a new PR.

**Pull request merged**